design/transaction-layer-refactoring.md

```diff
@@ -0,0 +1,151 @@
+# Transaction Layer Refactoring
+
+## Motivation
+
+At the very beginning, all transactional commands in TiKV share common procedures:
+
+1. Acquire latches
+2. Check constraints and generate modifications
+3. Apply modifications to the raft store
+4. Reply to RPC and release the latches
+
+After the latches and the snapshot are acquired, all the commands depend on nothing else, so we needn't care about passing dependencies in deeply.
+
+So the current implementation structure is natural and great: different commands are wrapped as enum variants and go through the whole process. In each step we match the type of command and do specific work.
+
+However, as more commands and optimizations are added, this is not always the case. Some commands breaks the procedure and adds dependencies.
+
+For example, the "pipelined pessimistic lock" replies before the raft store finishes replication, which is not following the common procedure. Also, when an `AcquirePessimisticLock` command meets a lock, it needs to wait for the lock being released in TiKV for a while. And when a lock is released, the awaiting `AcquirePessimisticLock` commands are notified. These are distinctive steps that are not shared by other commands. More dependencies (the lock manager and the deadlock detector) are also introduced.
+
+If more and more commands are not sharing the same procedure and dependencies, the current structure will have no benefit. Instead, the drawbacks become clearer.
+
+The code for a certain command is spread in too many files. Each function handles a part of work of various commands. People are hard to understand what happens about a single command.

```

Comment:
I find this is a very big problem. My plan to address it is to lean heavily on the `Command` type. I think that we can move all code for specific commands from process.rs and txn.rs into the Command implementation (and split up the command module so there is one submodule per command).